Extract the profiler to a new component

jelte opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | https://github.com/symfony/symfony/issues/11263 |
| License | MIT |
| Doc PR | https://github.com/symfony/symfony-docs/pull/5422 |

**This pull request is a replacement for https://github.com/symfony/symfony/pull/10374**

This pull request extracts the HTTP profiler from the HttpKernel component to a new dedicated component: Profiler.

The goals are to improve the general architecture and usability of the component as stated by @webmozart ( https://github.com/symfony/symfony/pull/10374#issuecomment-36616624 )
as it stands now no BC is broken as the original DataCollectors are still supported by the new Profiler.

New DataCollectors and Profiler no longer require a Request.
DataCollectors are should be able to collect data from other sources, which will prevent some ugly hacks that were previously required, they will still require to listen to events to collect the data, I don't see a way around this, but this doesn't seem a big issue.

If a DataCollector requires the Request, it should have the RequestStack and get it from there, the Responses can be added just like Controllers are added, by subscribing to the Kernel Events.
# Changes
- Profiler is no longer capable of searching or retrieving profiles, developers should use the ProfilerStorage directly for these functionalities.
- DataCollectors are no longer added to the Profile. A `ProfileDataInterface` object is added instead.
- DataCollectors no longer have a name.
- Saving the profile is independent if its indexes, this allows for storing other types of Profiles while using the same Profiler and ProfilerStorage. ( for example Console Commands )
- Many of the DataCollectors have been moved to their respective Component under their Profiler namespace.
# Profiling

``` php
$profiler = new Profiler($storageInterface);
$profiler->addCollector(new RequestDataCollector($requestStack));
$profiler->addCollector(new ExceptionDataCollector());
$profiler->addCollector(new FormDataCollector());

// Runtime data is collected
$profile = $profiler->profile();

$requestData = $profile->get('request');

// Late data is collected
$profiler->save($profile, array('url' => 'http://localhost/', 'ip' => '127.0.0.1', 'method' => 'GET', 'status_code' => 200, 'profile_type' => 'http');
```
# BC Breaks
## DataCollectors

**DataCollectors require `getCollectedData`.**

_solution:_
commented the method in the `DataCollectorInterface` and added a fallback when the method does not exists.
This will the use the `GenericProfileData` to encapsulate the DataCollector, which will trigger a deprecated notice.
## ProfilerStorages

**ProfilerStorages require `findBy` which replaces the `find` method.**

_solution:_
commented the method in the `ProfilerStorageInterface`. 
# Todo's
- [x] add additional tests to increase coverage.
- [x] add README.md
- [x] Test with Symfony demo
- [x] Convert Bundle & Bridge DataCollectors (if that does not break BC)
- [x] ConsoleProfiler
- [x] Move DataCollector to their related Component or Bridge
- [x] DoctrineDataCollector: Move collected data from DoctrineBundle to DoctrineBridge
- [x] Refactor FormDataCollector
- [x] Refactor DumpDataCollector as it has to many responsibilities
- [x] add tests for MysqlProfileStorage
- [x] Make ProfileData Serializable
- [x] Rename `DataCollectorInterface::collect` to `getCollectedData`
- [x] Remove `getName` from `DataCollectorInterface`
- [x] Remove storage methods (`load`, `find`, `purge`) from `Profiler` 
- [x] Remove `Collectors` from `Profile`
- [x] Remove `export` & `import` from `Profiler`
- [x] move all deprecated code to the (deprecated) subclass.
- [x] Check for BC breaks ( and fix where possible )
- [x] add phpdocs where missing
- [x] squash pull request
- [x] submit changes to the documentation

Special thanks to @stof & @dosten for helping me out on the initial pull request ( https://github.com/symfony/symfony/pull/14763 )

dosten commented on src/Symfony/Component/HttpKernel/DataCollector/DataCollectorInterface.php:
```diff
@@ -33,13 +36,4 @@
      * @api
      */
     public function collect(Request $request, Response $response, \Exception $exception = null);
```
Why you do not move this method definition to the new `DataCollectorInterface` and only keep this interface empty and extending the new one?

jelte replied:
There are 2 reason for this, for one the new DataCollectorInterface doesn't have a collect function any more.
RuntimeDataCollectorInterface replace the old DataCollectorInterface, as I noticed some if not most use the LateDataCollectorInterface and have no use for a collect method.
For that reason I have removed the collect method from DataCollectorInterface.

The other reason is that the new profiler does not require a Request and Response, but to maintain BC I had to leave this in.

dosten commented on src/Symfony/Component/HttpKernel/EventListener/ProfilerListener.php:
```diff
@@ -11,41 +11,34 @@
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @deprecated since 2.8, to be removed in 3.0. Use Symfony\Component\Profiler\EventListener\HttpProfileListener instead.
```
Typo: HttpProfileListener should be HttpProfilerListener

dosten commented on src/Symfony/Component/HttpKernel/Profiler/MongoDbProfilerStorage.php:
```diff
@@ -11,249 +11,42 @@
-    private function getProfile(array $data)
+/**
+ * Class MongoDbProfilerStorage
+ * @package Symfony\Component\HttpKernel\Profiler
```
We don't use the `@package` tag

mickaelandrieu commented on src/Symfony/Component/HttpKernel/composer.json:
```diff
@@ -20,6 +20,7 @@
         "symfony/event-dispatcher": "~2.5.9|~2.6,>=2.6.2|~3.0.0",
         "symfony/http-foundation": "~2.5,>=2.5.4|~3.0.0",
         "symfony/debug": "~2.6,>=2.6.2",
+        "symfony/profiler": "~2.7|~3.0.0",
```
this is probably
``
"symfony/profiler": "~2.8|~3.0.0"
``

mickaelandrieu commented on src/Symfony/Component/Profiler/DataCollector/DumpDataCollector.php:
```diff
@@ -0,0 +1,345 @@
+            KernelEvents::RESPONSE => array('onKernelResponse', -50)
+        );
+    }
+
```
extra line here

mickaelandrieu commented on src/Symfony/Component/Profiler/DataCollector/LoggerDataCollector.php:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+
```
extra line here

mickaelandrieu commented on src/Symfony/Component/Profiler/Profile.php:
```diff
@@ -0,0 +1,379 @@
+        return $this->getProfileData($name);
+    }
+
+    public function has($name)
```
missing  phpdocs here

mickaelandrieu commented on src/Symfony/Component/Profiler/ProfileData/ConfigData.php:
```diff
@@ -0,0 +1,185 @@
+ */
+class ConfigData extends AbstractProfileData
+{
+    public function getApplicationName()
```
don't forget phpdocs here as this is part of public api, same for ``getApplicationVersion``

jelte replied:
Many of the functions were copied from the original DataProfilers, but will go through everything and add appropriate phpdocs.

mickaelandrieu commented on src/Symfony/Component/Profiler/ProfileData/DumpData.php:
```diff
@@ -0,0 +1,65 @@
+        $this->dataCount = $dataCount;
+        $this->charset = $charset;
+    }
+
```
phpdocs.

phansys commented on src/Symfony/Bridge/Doctrine/Profiler/DoctrineDataCollector.php:
```diff
@@ -0,0 +1,143 @@
+        return 'db';
+    }
+
+    private function sanitizeQueries($connectionName, $queries)
```
2nd argument could be typehinted with ```array```.

mickaelandrieu commented on src/Symfony/Component/Profiler/ProfileData/DumpData.php:
```diff
@@ -0,0 +1,65 @@
+            if (method_exists($dump['data'], 'withMaxDepth')) {
+                $dumper->dump($dump['data']->withMaxDepth($maxDepthLimit)->withMaxItemsPerDepth($maxItemsPerDepth));
+            } else {
+                // getLimitedClone is @deprecated, to be removed in 3.0
```
if it's deprecated, does it need to "throw" a deprecated error?

phansys commented on src/Symfony/Component/Profiler/DataCollector/DumpDataCollector.php:
```diff
@@ -0,0 +1,345 @@
+use Symfony\Component\VarDumper\Dumper\CliDumper;
+use Symfony\Component\VarDumper\Dumper\HtmlDumper;
+use Symfony\Component\VarDumper\Dumper\DataDumperInterface;
+use Symfony\Component\Profiler\ProfileData\DumpData;
```
Please, check ```use``` order.

phansys commented on src/Symfony/Component/Profiler/DataCollector/RequestDataCollector.php:
```diff
@@ -0,0 +1,89 @@
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\HttpKernel\Event\FilterControllerEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
```
Check ```use``` order.

phansys commented on src/Symfony/Component/Profiler/DataCollector/RouterDataCollector.php:
```diff
@@ -0,0 +1,101 @@
+    {
+        return array(
+            KernelEvents::CONTROLLER => 'onKernelController',
+            KernelEvents::RESPONSE => array('onKernelException', -50)
```
Missing comma at end of multiline array item.

phansys commented on src/Symfony/Component/Profiler/DataCollector/TimeDataCollector.php:
```diff
@@ -0,0 +1,73 @@
+
+        if (null !== $this->kernel) {
+            $this->startTime = $this->kernel->getStartTime();
+        } else if (  null !== $request  ) {
```
Remove extra spaces.

phansys commented on src/Symfony/Component/Profiler/ProfileData/RequestData.php:
```diff
@@ -0,0 +1,312 @@
+ */
+class RequestData extends AbstractProfileData
+{
+
```
Remove extra line.

phansys commented on src/Symfony/Component/Profiler/Storage/AbstractMemcacheProfilerStorage.php:
```diff
@@ -0,0 +1,315 @@
+    /**
+     * Get name of index.
+     *
+     * @return string
```
```false``` isn't a string.

mickaelandrieu commented:
@jelte don't give up please, I need this component too and I will be happy to help you :+1:

jelte commented:
@mickaelandrieu I won't, a lot of the comments are issues coming from the original files, any others are good learning for me.

phansys commented:
I think you should use [PHP-CS-FIXER](https://github.com/FriendsOfPHP/PHP-CS-Fixer) to solve majority of your CS issues. It should leave a clear code to make the review easy.

jelte commented:
For some of the DataCollectors I have the feeling they do not belong in the Profiler:
- DumpDataCollector => VarDumper or DebugBundle, the DumpListener from HttpKernel should probably be moved as well.
- EventDataCollector => EventDispatcher
- LoggerDataCollector => MonologBridge, the DebugLoggerInterface should also be moved ( should be done carefully as this is used in the ExceptionListener )
- RequestDataCollector => HttpKernel
- RouteDataCollector => Routing

This is already done for the FormDataCollector and SecurityDataCollector.

Collectors that would remain in the Profiler are:
- ConfigDataCollector. I am considering splitting this one to ProjectDataCollector, SymfonyDataCollector and ConfigDataCollector, but this will have some impact on the WebProfilerBundle.
- MemoryDataCollector
- TimeDataCollector

mickaelandrieu commented:
@jelte totaly agree **but** I disagree to make new strong dependencies between the components.

Maybe FrameworkBundle is the better place for this collectors ?

jelte commented:
@mickaelandrieu I would keep them in their respective Bundle or Bridge if one exists, rather than dumping everything in the FrameworkBundle.

- DumpDataCollector => DebugBundle
- EventDataCollector => FrameworkBundle
- LoggerDataCollector => MonologBridge
- RequestDataCollector => FrameworkBundle
- RouteDataCollector => FrameworkBundle

mickaelandrieu commented:
@jelte agree :) /c @fabpot what do you think ?

stof commented:
Please don't move them into bundles. It would make the profiler unusable for people using the components (Silex for instance).

mickaelandrieu commented:
@stof humm, how about a ``symfony-collectors`` package then ?

jelte commented:
@stof I understand, that's why I suggested to keep them in the component they belong to (like what is done with the FormProfiler at the moment).
@mickaelandrieu this will just move the issue from profiler to symfony-collectors (creating a bundle that has dependency to everything).

aitboudad commented on src/Symfony/Bridge/Doctrine/Profiler/DoctrineDataCollector.php:
```diff
@@ -0,0 +1,143 @@
+    {
+        $this->registry = $registry;
+        $this->connections = $registry->getConnectionNames();
+        $this->managers = $registry->getManagerNames();
```
we could remove `managers` and `connections` properties, we don't need them here.

stof replied:
For the Doctrine bridge, the new collector might be a good time to move the extra data collected in DoctrineBundle to the bridge itself btw.

stof commented:
Well, once we have a dedicated component for the profiler, component profilers could indeed go in each component. this would be consistent with the fact that ContainerAwareDispatcher is in the EventDispatcher rather than the DI component for instance.
Each component with a profiling collector would then have an optional dependency on the Profiler component (optional because using the collector would be optional of course, and useful only for people using the profiler component)

mickaelandrieu commented:
@stof :+1:

jelte commented:
Travis-ci currently fails on https://travis-ci.org/symfony/symfony/jobs/67820438#L3999
> Failed to execute git clone --mirror 'git@github.com:jelte/symfony' '/home/travis/.composer/cache/vcs/git-github.com-jelte-symfony/'

Anyone got any idea how to fix this?

mickaelandrieu commented:
@jelte you have probably reach the limit for non authenticated users.

From your side, you need to set a public key into your github account, but from "travis" side I don't know what you can do ... /c @stof

jelte commented:
Think I'm finally finished, never expected this to become such a huge PR :sweat_smile:.

If everything worked out as planned, no BC should be broken. 
Next steps would be to provide PR for the DataCollectors in the separate bundles ( SwiftMailer, Doctrine, ... )

Any feedback now is more than welcome.

xabbuh commented on src/Symfony/Component/Profiler/composer.json:
```diff
@@ -0,0 +1,50 @@
+            "type": "git",
+            "url": "git@github.com:jelte/symfony"
+        }
+    ],
```
this looks wrong

jelte replied:
Thank for catching this. been searching for a wile. this will probably fix travis I guess.

webmozart commented:
First of all, WOW! Thanks for that awesome PR @jelte :) I think it will take several roundtrips to review this properly, but I'll try to start.

webmozart commented on src/Symfony/Component/Profiler/ProfileData/ProfileDataInterface.php:
```diff
@@ -0,0 +1,21 @@
+ *
+ * @author Jelte Steijaert <jelte@khepri.be>
+ */
+interface ProfileDataInterface
```
I think this interface should extend `\Serializable`. Any profile data should be serializable.

Nicofuma commented on src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php:
```diff
@@ -22,6 +22,8 @@
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @deprecated since 2.8, to be removed in 3.0. Use Symfony\Bridge\Twig\Profiler\DoctrineDataCollector instead.
```
> Symfony\Bridge\\**Doctrine**\Profiler\DoctrineDataCollector

webmozart commented on src/Symfony/Component/Profiler/DataCollector/RuntimeDataCollectorInterface.php:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return ProfileDataInterface
+     */
+    public function collect();
```
I think the name of this method is not quite appropriate. The method does not actually collect the data (the data is collected during the run of the application), but only return the collected data. Hence I think `getCollectedData()` would be more appropriate.

webmozart replied:
Also I think this should be moved to DataCollectorInterface. Instead of duplicating the method in this interface and LateDataCollectorInterface, I'd prefer something like `getPriority()` in DataCollectorInterface and to remove LateDataCollectorInterface as well.

jelte replied:
`getPriority()` might be confusing as it would not be defining a priority, but just an indication when it the data is collected. Either on `Profiler::profile()` or on `Profiler::save()`.

I will move and rename `collect()`.

webmozart commented on src/Symfony/Component/Profiler/DataCollector/DataCollectorInterface.php:
```diff
@@ -0,0 +1,36 @@
+     *
+     * @api
+     */
+    public function getName();
```
Why do Data collectors need names? What's the use case for accessing collectors by name? I think that this method should be moved to ProfileDataInterface.

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+        $this->logger = $logger;
+    }
+
+    public function generateToken()
```
Does this really need to be public?

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+     *
+     * @return ProfileInterface A Profile instance
+     */
+    public function load($token)
```
I think this method does not belong here. If you want to load a profile, use the storage directly.

jelte replied:
I don't disagree and the only reason to keep it might be **DX**, but using the `Storage` shouldn't be a big step.

but the storage will need to be publicly accessible then
```xml
<service id="profiler.storage" class="" public="false">
```

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+     *
+     * @return ProfileInterface A Profile instance
+     */
+    public function loadFromResponse(Response $response)
```
This method is coupled to HttpFoundation. I don't think it belongs here.

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+    /**
+     * Purges all data from the storage.
+     */
+    public function purge()
```
Again, this doesn't belong here. Use the storage directly.

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+     *
+     * @see http://php.net/manual/en/datetime.formats.php for the supported date/time formats
+     */
+    public function findBy(array $criteria, $limit, $start, $end)
```
Again, use the storage directly :)

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+     *
+     * @return string The exported data
+     */
+    public function export(ProfileInterface $profile)
```
Exporting and importing should not be the responsibility of the Profiler.

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+     *
+     * @return ProfileInterface|void
+     */
+    public function profileRequest(Request $request, Response $response)
```
I don't like this method because it's coupled to HttpFoundation. I think it should be removed. Setting the "X-Debug-Token" header is not the responsibility of a profiler.

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+     */
+    public function profile(ProfileInterface $profile, Request $request = null, Response $response = null, \Exception $exception = null)
+    {
+        foreach ($this->collectors as $collector) {
```
The check for `$this->enabled` is missing here.

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+                // we need to clone for sub-requests
+                $profile->addCollector(clone $collector);
+            } elseif ($collector instanceof DeprecatedDataCollectorInterface) {
+                @trigger_error(sprintf('%s implementing %s will is deprecated since version 2.8 and support for it will be removed in 3.0.', get_class($collector), 'Symfony\Component\HttpKernel\DataCollector\DataCollectorInterface'), E_USER_DEPRECATED);
```
What about moving deprecated functionality to the deprecated subclass?

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+                $profile->add($collector->getName(), $collector->collect());
+            } elseif ($collector instanceof LateDataCollectorInterface) {
+                // we need to clone for sub-requests
+                $profile->addCollector(clone $collector);
```
I'm not fond of storing collectors in the profile. How can we avoid that? The profile should contain data only and be serializable. Collectors are services, and as such are not serializable.

jelte replied:
I think we could avoid it by using the Collectors of the profiler in the `save()` method instead of the ones from the profile it self, this should effectively no longer require us to store a collector on the profile.
But I will need to check if the behaviour doesn't change as we would be remove the cloning of the collector.

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+     *
+     * @return array An array of collectors
+     */
+    public function all()
```
I don't think that it should be possible to access the collectors in the profiler. What's the use case?

jelte replied:
should indeed not be possible, can't find/see/think of any viable use case.

webmozart commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,387 @@
+     * @see http://php.net/manual/en/datetime.formats.php for the supported date/time formats
+     * @deprecated since 2.8 and will be removed in 3.0 use Profiler::findBy instead
+     */
+    public function find($ip, $url, $limit, $method, $start, $end)
```
I think you can move all deprecated code to the (deprecated) subclass.

webmozart commented on src/Symfony/Component/Profiler/ProfileInterface.php:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @param DataCollectorInterface[] $collectors
+     *
+     * @deprecated since 2.8. Will be removed in 3.1.
```
3.1?

jelte commented:
@webmozart  as some of the remarks relate to the dependency on HttpFoundation, I was taking a look at how we could remove most of this.

My idea came for how the profiles are stored in the `FileProfilerStorage`.
there the profile is basically split into 2 the `Profile` itself and the `IndexedProfile` containing the `url`, `ip`, etc.
We could potentially do the same in the `Profiler` and `<Http|Console>ProfilerListener`.

```php
class Profiler {
    /** @return Profile */
    public function profile();

    /** @return bool */
    public function save(IndexedProfileInterface $profile);
}
```
```php
class HttpProfilerListener {
    public function onKernelResponse(FilterResponseEvent $event) {
         ...
         $profile = $profiler->profile();
        
    }
}
```

jelte commented:
Although it's almost 4 weeks since the last addition, I'm still working on this but running into a few hiccups  which cause BC to break.

The introduction of `DataCollector::getCollectedData` and `ProfileData` in combination with removing `DataCollectors` from the `Profile` make it difficult to maintain BC, but I will find an elegant way (hopefully) to overcome this.

jelte commented:
This is ready to be reviewed,
Just need to fix the travis-ci (deps=min & deps=high) & appveyor builds.

pyrech commented:
Awesome work on this PR @jelte. Still planned for 2.8?

jelte commented:
@pyrech from my part yes, but it will depend on the code reviewing process and it seems travis-ci hates me as I got no clue what's wrong atm with it.

pyrech commented:
On appveyor, one failure on Process component seems unrelated but one failure is about the export command.

Regarding travis:
- the deps=high build fails on multiple tests (it seems that class `Symfony\Component\EventDispatcher\Profiler\EventDataCollector` is not found)
- deps=low build fails on SecurityBundle because no autoloader was found (may be a volatile bug that will be gone on a new run)
- hhvm build fails on the Console component and seems unrelated to this PR

Maybe @stof, @webmozart or @fabpot or any sf deciders can help here.

Status: Needs Review

xabbuh commented:
Is it intended that some of the deprecated classes do not trigger deprecation warnings?

Besides that the work done so far looks really good and I would be in favor of merging this for Symfony 2.8 (we still have two months to fix any issue we discover afterwards or make required tweaks).

mickaelandrieu commented on src/Symfony/Component/Profiler/README.md:
```diff
@@ -0,0 +1,45 @@
+==================
+
+Profiler collects information about each request made to your
+HttpKernel-based application and store them for later analysis.
```
except if I'm wrong, the Profiler component don't depend of the HttpKernel component, correct ?

jelte replied:
Problem at the moment is the ProfileListener, I moved it from the HttpKernel to prevent BC and not needing to rename it.

jelte replied:
The Profiler component is now completely independent from the HttpKernel component.

mickaelandrieu commented on src/Symfony/Component/Profiler/DataCollector/DataCollectorInterface.php:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @todo introduce in 3.0
+     */
+    //public function getCollectedData();
```
I'm not easy with commented code in code, you shoud move the function definition in ``@todo``

mickaelandrieu commented on src/Symfony/Component/Profiler/ProfileData/TimeData.php:
```diff
@@ -0,0 +1,130 @@
+ */
+
+namespace Symfony\Component\Profiler\ProfileData;
+use Symfony\Component\Stopwatch\Stopwatch;
```
is Stopwatch component an hard dependency ? In my opinion it shouldn't. In all cases, this need to be explicit in the composer.json file

jelte replied:
Think it's a optional dependency as you'd only need it when you enable the TimeDataCollector.

mickaelandrieu commented on src/Symfony/Component/Profiler/ProfileData/TokenAwareProfileDataInterface.php:
```diff
@@ -0,0 +1,17 @@
+<?php
+
```
header class missing here

mickaelandrieu commented on src/Symfony/Component/Profiler/Profiler.php:
```diff
@@ -0,0 +1,141 @@
+
+namespace Symfony\Component\Profiler;
+
+use Symfony\Component\HttpKernel\Profiler\GenericProfileData;
```
This GenericProfileData is here only for BC ? Too bad to create an hard dependency on HttpKernel component for a deprecated class :/

jelte replied:
It might be better to move it to the Profiler, but I don't think this is a big issue as it will be removed in 3.0.

mickaelandrieu replied:
Ok :)

mickaelandrieu commented on src/Symfony/Component/Profiler/Storage/AbstractProfilerStorage.php:
```diff
@@ -0,0 +1,85 @@
+<?php
+
```
header class missing here

mickaelandrieu commented on src/Symfony/Component/Profiler/Storage/AbstractProfilerStorage.php:
```diff
@@ -0,0 +1,85 @@
+namespace Symfony\Component\Profiler\Storage;
+
+use Symfony\Component\Profiler\Profile;
+
```
by the way, @author block is missing :)

mickaelandrieu commented on src/Symfony/Component/Profiler/Storage/ProfilerStorageInterface.php:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @todo introduce in 3.0
+     */
+    //public function findBy(array $criteria, $limit, $start = null, $end = null);
```
I'm not sure of the good solution here: commented code or explicit @todo ...

jelte replied:
This exists in other classes like that, so I just copied it.

mickaelandrieu commented on src/Symfony/Component/Profiler/Tests/DataCollector/ConfigDataCollectorTest.php:
```diff
@@ -0,0 +1,69 @@
+namespace Symfony\Component\Profiler\Tests\DataCollector;
+
+use Symfony\Component\Profiler\DataCollector\ConfigDataCollector;
+
```
@author is missing here

mickaelandrieu commented on src/Symfony/Component/Profiler/Tests/DataCollector/MemoryDataCollectorTest.php:
```diff
@@ -0,0 +1,58 @@
+
+use Symfony\Component\Profiler\DataCollector\MemoryDataCollector;
+use Symfony\Component\Profiler\ProfileData\MemoryData;
+
```
@author is missing here

mickaelandrieu commented on src/Symfony/Component/Profiler/Tests/DataCollector/TimeDataCollectorTest.php:
```diff
@@ -0,0 +1,57 @@
+use Symfony\Component\Profiler\DataCollector\TimeDataCollector;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Stopwatch\Stopwatch;
+
```
check all @uthor files :)

mickaelandrieu commented on src/Symfony/Component/Profiler/Tests/Mock/MockProfile.php:
```diff
@@ -0,0 +1,11 @@
+<?php
+
```
header file missing here (and author)

mickaelandrieu commented on src/Symfony/Component/Profiler/composer.json:
```diff
@@ -0,0 +1,43 @@
+    ],
+    "require": {
+        "php": ">=5.3.9",
+        "psr/log": "~1.0"
```
for now, http-kernel is an hard requirement, you need also the stop-watch component for the TimeDataCollector (can be a soft dependency)

mickaelandrieu commented on src/Symfony/Component/Profiler/composer.json:
```diff
@@ -0,0 +1,43 @@
+        "twig/twig": "~1.18"
+    },
+    "suggest": {
+        "symfony/http-kernel": ""
```
will be a suggest in 3.0 ;)

DavidBadura commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/form.html.twig:
```diff
@@ -693,6 +693,10 @@
         {{ tree.form_tree_details(childName, childData, forms_by_hash) }}
+=======
+        {{ _self.form_tree_details(childName, childData) }}
+>>>>>>> Stashed changes
```
unresolved conflict

Taluu commented:
As 2.8 reached feature freeze status, would this still be planned on 2.8 or only on 3.0 (or 2.9 if there is such a version) ? I'd love to see that one for 2.8 though.

ping @symfony/deciders

JHGitty commented:
Any updates?

jelte commented:
Updated this pull-request to the current 2.8 version,
Running the tests locally gives me the same errors as the 2.8 branch does.

Have refactored some of the code along the way and been able to make the ProfilerListener in HttpKernel BC, so no need for the Profiler component to have a dependency to the HttpKernel component.

As most of the ProfilerStorage's have become deprecated I've removed to refactored ones from the Profiler component.
And hopefully I didn't miss any of the updates to DataCollectors.

mickaelandrieu commented:
ping @symfony/deciders can someone take a look at this contribution? At least to say if this may have a chance to be merged or not?

jelte commented:
Seems I've resolved most of the build issues today,
the main issue was (still is) that travis takes the 2.8 version of the components and not the ones from my pull request which means that the different datacollectors that are now in their component or bridges aren't found. Any idea how I can resolve this?

The only builds that break atm are 5.3 and hhvm due to deprecation notices during the Functional tests of the Framework and Security Bundles.

Commit:
Extract the profiler to a new component

jelte commented:
At the moment, I have resolved all deprecation notices.
(Build still fails, but they are the same failures as the 2.8 branch)

Commit:
Require zendframework/zend-stdlib to fix dependency in ocramius/proxy-manager

jelte commented:
@stof @fabpot 

> ping @symfony/deciders can someone take a look at this contribution? At least to say if this may have a chance to be merged or not?

Tobion commented:
@jelte it most likely will not be part of 2.8/3.0 as we are trying to stabilize now and not make big changes. But considering the amount of work done here, I assume it's worth to merge this to 3.1 later.

JHGitty commented:
As I currently understand I need to enable debug on the AppKernel to use the time measure in the Profiler, right?

My idea was to enable the profiler in our "testing" environment in combination with enabled caching to check the performance of our application. The issue is that if I enable 'debug' in the AppKernel I will trigger a lot of other things which will result in non-caching (for example the CSS files are generated and not cached).

Will this new Profiler component fix this issue and will it provide a good solution for this?

jelte commented:
@Tobion Figured as much, just wanted confirmation, will keep this up-to-date as frequently as possible.

jelte commented:
@JHGitty You can already do this now by defining a second Profiler and ProfilerListener and only adding  the TimeCollector to it.
Just put it in a separate xml which is only loaded on your "testing" environment.

{code}
     <service id="test.profiler_listener" class="Symfony\Component\HttpKernel\EventListener\ProfilerListener">
            <tag name="kernel.event_subscriber" />
            <argument type="service" id="test.profiler" />
            <argument type="service" id="test.request_matcher"/>
            <argument>false</argument>
            <argument>false</argument>
            <argument type="service" id="request_stack" />
        </service>

        <service id="test.profiler" class="Symfony\Component\HttpKernel\Profiler\Profiler">
            <argument type="service" id="test.profiler.storage" />
            <call method="add">
                <argument type="service" id="data_collector.time"/>
            </call>
        </service>
{code}

Idea would be just to config what you'd need and nothing more. This would only get you the data being generated.

JHGitty commented:
@jelte Is there any documentation about this? I think somebody should add a documentation page about this.

stof commented:
@jelte you should probably use a separate time collector to avoid issues (collectors are stateful)

@JHGitty note that registering the time panel alone will not help much: most places registering data into the stopwatch component are controlled by the debug mode in FrameworkBundle: wrapping the event dispatcher into a traceable one to log timing info comes at a cost, which is why it is only done in debug mode (where we add more debugging info at the expense of performance).
The symfony profiler is a debugging profiler (to know what happens in your request), but not a timing performance profiler (for which the profiling overhead must be as small as possible). You should use a real performance profiler for that (for instance, https://blackfire.io which is free for the basic features).

Anyway, this should be discussed elsewhere than in this PR as it is unrelated (and the PR discussion should stay focused on reviewing the PR)

pyrech commented:
Now that 3.0 was released and 3.1 development is in progress, is this PR still considered by the core team? @jelte already did an awesome job but the PR now obviously needs to be rebased.

mickaelandrieu commented:
@pyrech we can't ask to @jelte to maintain a fork for years too. 

Before ask for rebase, maybe someone from the core team could say if this contribution could be accepted or not? 

Other way to do: fork once and for all, and make a bundle.

Taluu commented:
A bundle is not a solution... but making a real lib is (as all components can be used independently)

mickaelandrieu commented:
@Taluu sure I saw this bundle as an implementation of this fork (which is already a library/component)

jelte commented:
Just to give everyone a little update,
I just started to rebase this on the current 3.0, I will update as soon as possible.

Shine-neko commented:
@jelte  ?

Koc commented:
Fixes #21141

Nek- commented on src/Symfony/Bridge/Doctrine/DataCollector/DoctrineDataCollector.php:
```diff
@@ -22,6 +22,8 @@
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @deprecated since 2.8, to be removed in 3.0. Use Symfony\Bridge\Doctrine\Profiler\DoctrineDataCollector instead.
```
in 4.0 ?

Nek- commented on src/Symfony/Bridge/Twig/DataCollector/TwigDataCollector.php:
```diff
@@ -20,6 +20,8 @@
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @deprecated since 2.8, to be removed in 3.0. Use Symfony\Bridge\Twig\Profiler\TwigDataCollector instead.
```
4.0 ?

Shine-neko commented:
@fabpot planned for 4.0 ?

fabpot commented:
I'm closing this one. I'm not convinced that we should have a decoupled profiler. The profiler does the job as is and the fact it's tied to Request/Response is part of the design that I don't want to change. I know that it took time for me to make that decision, but after all this time, I still think that's not something that needs to be done. Thanks for working on this.

Wirone commented:
This would be soooooooo helpful and I am so disappointed by @fabpot's decision :( Some applications doesn't even have HTTP context, they're 100% CLI and profiling them is a must have. I don't understand why it was rejected, at this point it would be probably crucial..

Feel bad for you, @jelte :(
